### PR TITLE
Setup Google Auth with Expo, API Middleware, API Tests Support Auth

### LINKED ISSUE #19: Setup Google Auth with Expo
**Is your feature request related to a problem? Please describe.**
Users need a way to login to the app.

**Describe the solution you'd like**

Expo can be used to login to many popular providers on iOS, Android, and web! Most of these guides utilize the pure JS [AuthSession](https://docs.expo.dev/versions/latest/sdk/auth-session/). We need to setup Google Auth for this project. This issue doesn't focus on creating the page for signup/login, but instead the client-side logic for facilitating this process. 

This issue will need to cover the following objectives:

- Setup GCP OAuth credentials
- Expo code for connecting to Google Auth
- Handle updating application state when user logins in
- Call the necessary endpoints for creating a user

**Resources**
- https://inaguirre.medium.com/react-native-login-with-google-quick-guide-fe351e464752
- https://docs.expo.dev/guides/authentication/

### PR DESCRIPTION
**Pull request by ashayp22**
## Status:

🚀 

## Description

Implemented all steps of the Auth Process ([see visual diagram](https://docs.google.com/drawings/d/1vslY8hguiTLQ3uj5D9EDbllL_7cc6oANSnu-CauZp18/edit)).

- Added Google Auth to the client-side with Expo.
- Saving ID tokens on the client-side with [Async Storage](https://react-native-async-storage.github.io/async-storage/) and Redux Middleware.
- Updated React Navigation to take into account users logging in.
- Built basic user auth middleware that verifies user with Google Auth on the backend and checks if they exist in MongoDB.
- Added support for auth middleware to API tests. 

Fixes #19 

## Todos

- [x] Add call to `POST '/user/' `(adds a user to MongoDB) to client.

### CODE REVIEW

File: api/src/middleware/authentication.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ */
+module.exports.requireAuthentication = async (req, res, next) => {
+  try {
+    // Validate user using Google Auth ID Token
+    const user = await getUserFromRequest(req);
+    if (!user) {
+      sendResponse(res, 401, ERR_IMPROPER_ID_TOKEN);
+    } else {
+      // Checks if user exists in MongoDB
+      const userInMongo = await models.User.findOne({ authID: user.sub }); // user.sub returns the user's Google Account unique ID
```
Comment (ashayp22) [inline api/src/middleware/authentication.js:28]: use .exists, restructure req.user

File: api/test/utils/auth.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports.withAuthentication = (request) =>
+  request.set({
+    authorization: 'Bearer authidtoken',
```
Comment (ayogik) [inline api/test/utils/auth.js:3]: What does this do (with 'authidtoken' being a literal string)?

File: client/app.config.js [hunk 3]
```diff
@@ -3,6 +3,10 @@ export default ({ config }) => {
     ...config,
     extra: {
       developmentApiPort: '3000',
+      iosClientId:
+        '558041423211-hc41ckr5muh3l09rsqqrs4nkaspltrba.apps.googleusercontent.com',
```
Comment (ayogik) [inline client/app.config.js:7]: how sensitive are these values?

File: client/src/api/api.js [hunk 4]
```diff
@@ -1,11 +1,18 @@
 import instance from './axios-config'
 
 export const getSampleHome = async () => {
-  const requestString = '/home'
+  const requestString = '/user'
   const res = await instance.get(requestString)
 
   if (!res?.data?.success) throw new Error(res?.data?.message)
```
Comment (ayogik) [inline client/src/api/api.js:7]: what's our stance on having both the condition and the statement in the same line?

File: client/src/pages/Details/Details.js [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import React from 'react'
 import PropTypes from 'prop-types'
-import { StyleSheet, Text, View, StatusBar } from 'react-native'
```
Comment (ayogik) [inline client/src/pages/Details/Details.js:3]: Reason for this?

File: client/src/api/api.js [hunk 6]
```diff
@@ -1,11 +1,18 @@
 import instance from './axios-config'
 
 export const getSampleHome = async () => {
```
Comment (ayogik) [inline client/src/api/api.js:3]: What does this do?

File: client/src/redux/store.js [hunk 7]
```diff
@@ -34,7 +35,9 @@ const defaultMiddleware = getDefaultMiddleware({
 const store = configureStore({
   reducer: rootReducer,
   // eslint-disable-next-line no-undef
-  middleware: __DEV__ ? defaultMiddleware.concat(logger) : defaultMiddleware,
+  middleware: __DEV__
+    ? defaultMiddleware.concat(logger).concat(saveAuthToken)
```
Comment (ayogik) [inline client/src/redux/store.js:39]: What is the logger doing?

File: client/src/utils/constants.js [hunk 8]
```diff
@@ -0,0 +1,2 @@
+export const ASYNC_STORAGE_ID_TOKEN_KEY = 'userToken'
+export const RANDOM_CONSTANT = true
```
Comment (ayogik) [inline client/src/utils/constants.js:2]: What is this?

**Review by ayogik — COMMENTED** [review]
Looks good, some points for clarification. Also, there appear to be elements of Nayonika + Jessica's and Ellie + Ishaan's PRs in here. How can we keep things more discrete in the future?

File: api/src/api/user.js [hunk 9]
```diff
@@ -29,4 +34,12 @@ router.post(
   }),
 );
 
+router.get(
+  '/',
+  requireAuthentication,
+  errorWrap(async (req, res) =>
+    sendResponse(res, 200, 'test data', `test data from api${Math.random()}`),
```
[the remark this answers is not in the document]
Comment (ashayp22) [reply api/src/api/user.js:41]: Just a random endpoint used for testing the requireAuthentication middleware. I can remove it.

[the remark this answers is not in the document]
Comment (ashayp22) [reply api/src/api/user.js:41]: Actually, I'm going to leave it as an example for users.

File: client/app.config.js (hunk 3 above)
Comment (ashayp22) [reply client/app.config.js:7]: oop these shouldn't be pushed

**Pull request merged**